File: src/core/safety/auto-approval.ts
/**
 * Auto-approval heuristic for dynamic tool calls (symphony #9, G5.G.6).
 *
 * Many tool calls during agent runs are cheap and obviously safe:
 * `Read`, `Glob`, `Grep`, `ls`, `pwd`, `git status`. Prompting the user
 * for each one is noise that trains them to click "allow" reflexively
 * (the worst possible muscle memory). This classifier identifies the
 * narrow set of "definitely safe" calls and approves them silently;
 * everything else falls through to the normal permission prompt path.
 *
 * Default-deny: if we cannot positively confirm a call is in the safe
 * subset, we return `prompt`. There is NO heuristic for "probably ok"
 * — the cost of a false positive is too high (data loss, network
 * egress with secrets, unintended writes).
 *
 * Compound bash → never auto-approve (INV-15 protection). Even if every
 * leg is in the safe-list, granting `allow` would bypass the scope-clamp
 * in `permissionService.check` that restricts compound bash to
 * scope=["once"]. We push to `prompt` so the user can confirm and
 * scope-clamp is applied. Detection uses the canonical `splitCompound`
 * helper (handles `&&`, `||`, `;`, `|`, `&`, `$()`, backticks, `<()`,
 * newlines, and quoting) — never a literal regex, which would miss
 * substitution and quoted-string edge cases.
 */
import { splitCompound } from "./bashCompoundAnalyzer";

export interface ToolCall {
	/** Tool name, e.g. "Read", "Bash", "Edit". */
	tool: string;
	/** Tool input. Shape is per-tool; we only inspect what we recognize. */
	input?: Record<string, unknown>;
}

export interface AutoApprovalDecision {
	decision: "allow" | "prompt" | "deny";
	reason: string;
}

// Tools that NEVER mutate state and NEVER egress network.
const READ_ONLY_TOOLS: ReadonlySet<string> = new Set([
	"Read",
	"Glob",
	"Grep",
	"LS",
	"NotebookRead",
]);

// First-word commands that are safe by themselves with any arg combination.
// Anything not on this list defaults to prompt.
const SAFE_BASH_COMMANDS: ReadonlySet<string> = new Set([
	"ls",
	"pwd",
	"cd",
	"echo",
	"cat",
	"head",
	"tail",
	"wc",
	"file",
	"stat",
	"which",
	"whoami",
	"hostname",
	"uname",
	"date",
	"env",
	"true",
	"false",
]);

// `git` is allowed only with read-only subcommands.
const SAFE_GIT_SUBCOMMANDS: ReadonlySet<string> = new Set([
	"status",
	"log",
	"diff",
	"show",
	"branch",
	"remote",
	"config",
	"rev-parse",
	"ls-files",
	"ls-tree",
	"blame",
	"describe",
	"reflog",
	"stash",
	"tag",
	"shortlog",
	"name-rev",
	"cat-file",
]);

// Destructive / remote-mutating tokens that ALWAYS push to prompt.
const HARD_DENY_TOKENS = [
	"rm",
	"sudo",
	"dd",
	"mkfs",
	"reboot",
	"shutdown",
	"halt",
	"poweroff",
	"chmod",
	"chown",
	"mv",
	"cp",
	"curl",
	"wget",
	"nc",
	"ncat",
	"telnet",
	"ssh",
	"scp",
	"rsync",
];

function classifyBashLeg(cmd: string): AutoApprovalDecision {
	const tokens = cmd.trim().split(/\s+/);
	if (tokens.length === 0 || tokens[0] === "") {
		return { decision: "prompt", reason: "empty command, not auto-approvable" };
	}
	const head = tokens[0];

	if (HARD_DENY_TOKENS.includes(head)) {
		return {
			decision: "prompt",
			reason: `command starts with "${head}", a destructive or network-mutating token`,
		};
	}

	if (SAFE_BASH_COMMANDS.has(head)) {
		return { decision: "allow", reason: `"${head}" is read-only / inert` };
	}

	if (head === "git") {
		const sub = tokens[1] ?? "";
		if (SAFE_GIT_SUBCOMMANDS.has(sub)) {
			return { decision: "allow", reason: `git ${sub} is read-only` };
		}
		return {
			decision: "prompt",
			reason: `git subcommand "${sub}" is not in the safe-list`,
		};
	}

	return {
		decision: "prompt",
		reason: `command "${head}" is not in the safe-list (default-deny)`,
	};
}

/**
 * Classify a tool call for auto-approval. See module header for the
 * default-deny posture and the composition rule.
 */
export function classifyToolForAutoApproval(call: ToolCall): AutoApprovalDecision {
	if (READ_ONLY_TOOLS.has(call.tool)) {
		return { decision: "allow", reason: `${call.tool} is a read-only tool` };
	}

	if (call.tool === "Bash") {
		const command = String(call.input?.command ?? "").trim();
		if (!command) {
			return { decision: "prompt", reason: "Bash call has no command" };
		}
		// INV-15: compound bash NEVER auto-approves, even if every leg is in
		// the safe-list. Auto-approval returns `allow` which short-circuits
		// the scope-clamp in `permissionService.check`; falling through to
		// `prompt` lets the clamp restrict scopes to ["once"].
		const legs = splitCompound(command);
		if (legs.length > 1) {
			return {
				decision: "prompt",
				reason: "compound bash skipped from auto-approval (INV-15 scope-clamp)",
			};
		}
		if (legs.length === 0) {
			return { decision: "prompt", reason: "Bash call is empty after splitting" };
		}
		const decision = classifyBashLeg(legs[0]);
		if (decision.decision !== "allow") {
			return decision;
		}
		return decision;
	}

	// Anything mutating or unknown: prompt.
	return {
		decision: "prompt",
		reason: `tool "${call.tool}" is mutating or unknown`,
	};
}

File: crates/apohara-safety/src/auto_approval.rs
//! Auto-approval heuristic for dynamic tool calls — ports
//! `src/core/safety/auto-approval.ts` (symphony #9, G5.G.6).
//!
//! Default-deny: only return `Allow` when we positively confirm a call
//! is in the safe subset. Everything else returns `Prompt`. There is no
//! "probably ok" heuristic — false positives are too expensive (data
//! loss / network egress with secrets).

use crate::bash_compound::split_compound;
use crate::patterns::ToolInvocation;
use std::collections::HashSet;
use std::sync::OnceLock;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AutoApprovalDecision {
    Allow { reason: String },
    Prompt { reason: String },
    Deny { reason: String },
}

impl AutoApprovalDecision {
    pub fn is_allow(&self) -> bool {
        matches!(self, AutoApprovalDecision::Allow { .. })
    }
}

fn read_only_tools() -> &'static HashSet<&'static str> {
    static SET: OnceLock<HashSet<&'static str>> = OnceLock::new();
    SET.get_or_init(|| {
        ["Read", "Glob", "Grep", "LS", "NotebookRead"]
            .into_iter()
            .collect()
    })
}

fn safe_bash_commands() -> &'static HashSet<&'static str> {
    static SET: OnceLock<HashSet<&'static str>> = OnceLock::new();
    SET.get_or_init(|| {
        [
            "ls", "pwd", "cd", "echo", "cat", "head", "tail", "wc", "file", "stat", "which",
            "whoami", "hostname", "uname", "date", "env", "true", "false",
        ]
        .into_iter()
        .collect()
    })
}

fn safe_git_subcommands() -> &'static HashSet<&'static str> {
    static SET: OnceLock<HashSet<&'static str>> = OnceLock::new();
    SET.get_or_init(|| {
        [
            "status", "log", "diff", "show", "branch", "remote", "config", "rev-parse",
            "ls-files", "ls-tree", "blame", "describe", "reflog", "stash", "tag", "shortlog",
            "name-rev", "cat-file",
        ]
        .into_iter()
        .collect()
    })
}

fn hard_deny_tokens() -> &'static HashSet<&'static str> {
    static SET: OnceLock<HashSet<&'static str>> = OnceLock::new();
    SET.get_or_init(|| {
        [
            "rm", "sudo", "dd", "mkfs", "reboot", "shutdown", "halt", "poweroff", "chmod",
            "chown", "mv", "cp", "curl", "wget", "nc", "ncat", "telnet", "ssh", "scp", "rsync",
        ]
        .into_iter()
        .collect()
    })
}

fn classify_bash_leg(cmd: &str) -> AutoApprovalDecision {
    let tokens: Vec<&str> = cmd.split_whitespace().collect();
    if tokens.is_empty() {
        return AutoApprovalDecision::Prompt {
            reason: "empty command, not auto-approvable".to_string(),
        };
    }
    let head = tokens[0];

    if hard_deny_tokens().contains(head) {
        return AutoApprovalDecision::Prompt {
            reason: format!(
                "command starts with \"{head}\", a destructive or network-mutating token"
            ),
        };
    }

    if safe_bash_commands().contains(head) {
        return AutoApprovalDecision::Allow {
            reason: format!("\"{head}\" is read-only / inert"),
        };
    }

    if head == "git" {
        let sub = tokens.get(1).copied().unwrap_or("");
        if safe_git_subcommands().contains(sub) {
            return AutoApprovalDecision::Allow {
                reason: format!("git {sub} is read-only"),
            };
        }
        return AutoApprovalDecision::Prompt {
            reason: format!("git subcommand \"{sub}\" is not in the safe-list"),
        };
    }

    AutoApprovalDecision::Prompt {
        reason: format!("command \"{head}\" is not in the safe-list (default-deny)"),
    }
}

/// Classify a tool call. Compound bash NEVER auto-approves (INV-bash-scope
/// scope-clamp protection — see `permission_service` for context).
pub fn classify_tool_for_auto_approval(inv: &ToolInvocation) -> AutoApprovalDecision {
    if read_only_tools().contains(inv.tool.as_str()) {
        return AutoApprovalDecision::Allow {
            reason: format!("{} is a read-only tool", inv.tool),
        };
    }

    if inv.tool == "Bash" {
        let command = inv
            .input
            .get("command")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim();
        if command.is_empty() {
            return AutoApprovalDecision::Prompt {
                reason: "Bash call has no command".to_string(),
            };
        }
        // INV-bash-scope: compound never auto-approves even if every leg
        // is safe. `allow` would short-circuit the scope-clamp in
        // `permission_service::check`.
        let legs = split_compound(command);
        if legs.len() > 1 {
            return AutoApprovalDecision::Prompt {
                reason: "compound bash skipped from auto-approval (INV-bash-scope clamp)"
                    .to_string(),
            };
        }
        if legs.is_empty() {
            return AutoApprovalDecision::Prompt {
                reason: "Bash call is empty after splitting".to_string(),
            };
        }
        return classify_bash_leg(&legs[0]);
    }

    AutoApprovalDecision::Prompt {
        reason: format!("tool \"{}\" is mutating or unknown", inv.tool),
    }
}
